# (PART) Day 2 {-}

```{r setup5, include=FALSE} 
library(dada2)
library(tidyverse)
library(phyloseq)

path <- "~/Desktop/microbiomeworkshop2022/rawdata"

fs <- sort(list.files(path, pattern = "_R1_001.fastq.gz", full.names = TRUE))
rs <- sort(list.files(path, pattern = "_R2_001.fastq.gz", full.names = TRUE))
sample.names <- sapply(strsplit(fs, "_R"), `[`, 1)


ffilt <- file.path(path, "filtered", paste0(basename(sample.names), "_F_filtered.fastq.gz"))
rfilt <- file.path(path, "filtered", paste0(basename(sample.names), "_R_filtered.fastq.gz"))

path <- "~/Desktop/microbiomeworkshop2022/results"
out <- readRDS(file.path(path, "filt_out.rds"))
```

# Denoise

The `DADA2` algorithm makes use of a parametric error model (`err`) and every amplicon dataset has a different set of error rates. The `learnErrors` method learns this error model from the data, by alternating estimation of the error rates and inference of sample composition until they converge on a jointly consistent solution. As in many machine-learning problems, the algorithm must begin with an initial guess, for which the maximum possible error rates in this data are used (the error rates if only the most abundant sequence is correct and all the rest are errors).

## Learn the Error Rates

Note: this is the most computationally intensive part of the `dada2` pipeline. Depending on your computer, this could take a while.

```{r}
errF <- learnErrors(ffilt, multithread = TRUE)
errR <- learnErrors(rfilt, multithread = TRUE)
```

Visualize the estimated error rates:

```{r, warning = FALSE}
plotErrors(errF, nominalQ = TRUE)
```

The error rates for each possible transition (A→C, A→G, …) are shown. Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score. Here the estimated error rates (black line) are a good fit to the observed rates (points), and the error rates drop with increased quality as expected. Everything looks reasonable and we proceed with confidence.

## Dereplicate

```{r}
derep_forward <- derepFastq(ffilt)
derep_reverse <- derepFastq(rfilt)
```

## Sample Inference

We are now ready to apply the core sample inference algorithm to the filtered and trimmed sequence data.

```{r}
dadaFs <- dada(derep_forward, err = errF, multithread = TRUE)
dadaRs <- dada(derep_reverse, err = errR, multithread = TRUE)
```

Inspecting the returned dada-class object:

```{r}
dadaFs[[1]]
```

There is much more to the dada-class return object than this (see help("dada-class") for some info), including multiple diagnostics about the quality of each denoised sequence variant, but that is beyond the scope of an introductory tutorial.

## Merge reads

We now merge the forward and reverse reads together to obtain the full denoised sequences. Merging is performed by aligning the denoised forward reads with the reverse-complement of the corresponding denoised reverse reads, and then constructing the merged “contig” sequences. By default, merged sequences are only output if the forward and reverse reads overlap by at least 12 bases, and are identical to each other in the overlap region (but these conditions can be changed via function arguments).

```{r}
merged <- mergePairs(dadaFs, ffilt, dadaRs, rfilt, verbose = TRUE)
# Inspect the merger data.frame from the first sample
head(merged[[1]])
```

The `merged` object is a list of data.frames from each sample. Each data.frame contains the merged `$sequence`, its `$abundance`, and the indices of the `$forward` and `$reverse` sequence variants that were merged. Paired reads that did not exactly overlap were removed by `mergePairs`, further reducing spurious output.

**Considerations**: Most of your reads should successully merge.If that is not the case upstream parameters may need to be revisited: Did you trim away the overlap between your reads?

## Construct sequence table

We can now construct an amplicon sequence variant table (ASV) table, a higher-resolution version of the OTU table produced by traditional methods.

```{r}
seqtab <- makeSequenceTable(merged)
dim(seqtab)

rownames(seqtab) <- basename(sample.names)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

The sequence table is a matrix with rows corresponding to (and named by) the samples, and columns corresponding to (and named by) the sequence variants.

### Exercise

Are these sequences the length that you expect? How are the reads concentrated in the merged sequence lengths? Plot the distribution of sequence lengths to explore. Something like the plot shown below.

```{r, echo = FALSE}
readsbyseqlen <- tapply(colSums(seqtab), nchar(colnames(seqtab)),sum)
plot(as.integer(names(readsbyseqlen)),readsbyseqlen, xlab = "Merged length", ylab = "Total reads")
```

### Exercise

Sequences that are much longer or shorter than expected may be the result of non-specific priming. Remove the non-target-length sequences from your sequence table. This is analogous to “cutting a band” in-silico to get amplicons of the targeted length.

```{r, echo = FALSE}
seqtab <- seqtab[,nchar(colnames(seqtab)) %in% 400:450]
```

After size selection:

```{r}
dim(seqtab)
table(nchar(getSequences(seqtab)))
```

## Remove Chimeras

The core dada method corrects substitution and indel errors, but chimeras remain. Fortunately, the accuracy of sequence variants after denoising makes identifying chimeric ASVs simpler than when dealing with fuzzy OTUs. Chimeric sequences are identified if they can be exactly reconstructed by combining a left-segment and a right-segment from two more abundant “parent” sequences.

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, 
                                    method="consensus", 
                                    multithread=TRUE, 
                                    verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

The frequency of chimeric sequences varies substantially from dataset to dataset, and depends on on factors including experimental procedures and sample complexity.

**Consideration**: Most of your reads should remain after chimera removal (it is not uncommon for a majority of sequence variants to be removed though). If most of your reads were removed as chimeric, upstream processing may need to be revisited. In almost all cases this is caused by primer sequences with ambiguous nucleotides that were not removed prior to beginning the DADA2 pipeline.
 
## Track reads through the pipeline

As a final check of our progress, we’ll look at the number of reads that made it through each step in the pipeline:

```{r}
getN <- function(x) sum (getUniques(x))
summary_tab <- data.frame(row.names=basename(sample.names), 
                          Input=out[,1], 
                          Filtered=out[,2],
                          Denoised=sapply(dadaFs, getN), 
                          Merged=sapply(merged, getN), 
                          Non.Chimeric=rowSums(seqtab.nochim), 
                          Total.Perc.Remaining = round(rowSums(seqtab.nochim)/out[,1]*100 ,1))


## Write this table to output
write.table(summary_tab, file.path(path, "reads_tracked.txt"))
```

### Exercise

Make a plot to visualize the results in `reads_tracked.txt`. Something like this:

```{r, echo = FALSE}
summary_tab$Sample <- rownames(summary_tab) 
summary_tab <- summary_tab %>% separate(Sample, c("Sample", "temp"), sep = "_S") 
summary_tab$Sample <- factor(summary_tab$Sample, levels = summary_tab$Sample[order(summary_tab$Non.Chimeric)])
summary_tab_long <- summary_tab %>% gather("QC.Step", "Reads", Input:Non.Chimeric)
summary_tab_long$QC.Step <- factor(summary_tab_long$QC.Step, levels = c("Input", "Filtered", "Denoised", "Merged", "Non.Chimeric"))


gg <- ggplot(summary_tab_long, aes(x = Sample, y = Reads, color = QC.Step)) +
	geom_point(size = 2) +
	scale_color_manual(values = rainbow(5, v = 0.8)) +
	theme_bw() +
	theme(axis.text.x = element_text(angle = 90))
gg
```

# Assign Taxonomy

We will use the function `assignTaxonony` to assign taxonomy to the sequence variants. This function takes as input a set of sequences to be classified and a training set of reference sequences with known taxonomy, and outputs taxonomic assignment with at least `minBoot` bootstrap confidence. 

## Download the database

Download the following databases from the [Silva website](https://zenodo.org/record/4587955#.YtYY2C2cZo4):

- silva_nr99_v138.1_train_set.fa.gz
- silva_species_assignment_v138.1.fa.gz

```{r}
path_to_refs <- "~/Desktop/dropbox_temp/bcch/reference_files"
taxa <- assignTaxonomy(seqtab.nochim, file.path(path_to_refs, "silva_nr99_v138_train_set.fa"), multithread=TRUE)

```

The dada2 package also implements a method to make species level assignments based on exact matching between ASVs and sequenced reference strains. Recent analysis suggests that exact matching (or 100% identity) is the only appropriate way to assign species to 16S gene fragments. Currently, species-assignment training fastas are available for the Silva and RDP 16S databases. To follow the optional species addition step, download the silva_species_assignment_v132.fa.gz file, and place it in the directory with the fastq files.

```{r}
taxa <- addSpecies(taxa, file.path(path_to_refs, "silva_species_assignment_v138.fa"))

```

### Exercise

Look at the usage documentation for the function `addSpecies`, in particular the option `allowMultiple`. Considering your research question, do you want to set this parameter to `TRUE` or `FALSE`? 

Inspect the taxonomic assignments:

```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

Finally, we've reached the end of the `dada2` pipeline. But first, we need to save the two files we've generated:

```{r}
path <- "~/Desktop/microbiomeworkshop2022/results"
saveRDS(seqtab.nochim, file.path(path, "seqtab_final.rds"))
saveRDS(taxa, file.path(path, "taxa_final.rds"))
```











