<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5 Denoise | 16s Microbiome Analysis Workshop</title>
  <meta name="description" content="This workbook contains workshop material for all three days." />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="5 Denoise | 16s Microbiome Analysis Workshop" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This workbook contains workshop material for all three days." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5 Denoise | 16s Microbiome Analysis Workshop" />
  
  <meta name="twitter:description" content="This workbook contains workshop material for all three days." />
  

<meta name="author" content="Alana Schick" />


<meta name="date" content="2022-07-18" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="filter.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">16S Analysis Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#workshop-schedule"><i class="fa fa-check"></i>Workshop Schedule</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-1"><i class="fa fa-check"></i>Day 1</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-2"><i class="fa fa-check"></i>Day 2</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-3"><i class="fa fa-check"></i>Day 3</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#important-links"><i class="fa fa-check"></i>Important links</a></li>
</ul></li>
<li class="part"><span><b>I Day 1</b></span></li>
<li class="chapter" data-level="1" data-path="pre-workshop.html"><a href="pre-workshop.html"><i class="fa fa-check"></i><b>1</b> Pre-workshop</a>
<ul>
<li class="chapter" data-level="1.1" data-path="pre-workshop.html"><a href="pre-workshop.html#packages"><i class="fa fa-check"></i><b>1.1</b> Packages</a></li>
<li class="chapter" data-level="1.2" data-path="pre-workshop.html"><a href="pre-workshop.html#data-download"><i class="fa fa-check"></i><b>1.2</b> Data download</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="introduction.html"><a href="introduction.html#basics-and-background"><i class="fa fa-check"></i><b>2.1</b> Basics and Background</a></li>
<li class="chapter" data-level="2.2" data-path="introduction.html"><a href="introduction.html#data"><i class="fa fa-check"></i><b>2.2</b> Data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="introduction.html"><a href="introduction.html#exercise"><i class="fa fa-check"></i><b>2.2.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="removing-primers.html"><a href="removing-primers.html"><i class="fa fa-check"></i><b>3</b> Removing Primers</a>
<ul>
<li class="chapter" data-level="3.1" data-path="removing-primers.html"><a href="removing-primers.html#identify-primers"><i class="fa fa-check"></i><b>3.1</b> Identify Primers</a></li>
<li class="chapter" data-level="3.2" data-path="removing-primers.html"><a href="removing-primers.html#remove-primers"><i class="fa fa-check"></i><b>3.2</b> Remove Primers</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="filter.html"><a href="filter.html"><i class="fa fa-check"></i><b>4</b> Filter</a>
<ul>
<li class="chapter" data-level="4.1" data-path="filter.html"><a href="filter.html#inspect"><i class="fa fa-check"></i><b>4.1</b> Inspect</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="filter.html"><a href="filter.html#exercise-1"><i class="fa fa-check"></i><b>4.1.1</b> Exercise</a></li>
<li class="chapter" data-level="4.1.2" data-path="filter.html"><a href="filter.html#exercise-2"><i class="fa fa-check"></i><b>4.1.2</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="filter.html"><a href="filter.html#filter-and-trim"><i class="fa fa-check"></i><b>4.2</b> Filter and Trim</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="filter.html"><a href="filter.html#exercise-3"><i class="fa fa-check"></i><b>4.2.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="denoise.html"><a href="denoise.html"><i class="fa fa-check"></i><b>5</b> Denoise</a>
<ul>
<li class="chapter" data-level="5.1" data-path="denoise.html"><a href="denoise.html#learn-the-error-rates"><i class="fa fa-check"></i><b>5.1</b> Learn the Error Rates</a></li>
<li class="chapter" data-level="5.2" data-path="denoise.html"><a href="denoise.html#dereplicate"><i class="fa fa-check"></i><b>5.2</b> Dereplicate</a></li>
<li class="chapter" data-level="5.3" data-path="denoise.html"><a href="denoise.html#sample-inference"><i class="fa fa-check"></i><b>5.3</b> Sample Inference</a></li>
<li class="chapter" data-level="5.4" data-path="denoise.html"><a href="denoise.html#merge-reads"><i class="fa fa-check"></i><b>5.4</b> Merge reads</a></li>
<li class="chapter" data-level="5.5" data-path="denoise.html"><a href="denoise.html#construct-sequence-table"><i class="fa fa-check"></i><b>5.5</b> Construct sequence table</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="denoise.html"><a href="denoise.html#exercise-4"><i class="fa fa-check"></i><b>5.5.1</b> Exercise</a></li>
<li class="chapter" data-level="5.5.2" data-path="denoise.html"><a href="denoise.html#exercise-5"><i class="fa fa-check"></i><b>5.5.2</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">16s Microbiome Analysis Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="denoise" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">5</span> Denoise<a href="denoise.html#denoise" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The <code>DADA2</code> algorithm makes use of a parametric error model (<code>err</code>) and every amplicon dataset has a different set of error rates. The <code>learnErrors</code> method learns this error model from the data, by alternating estimation of the error rates and inference of sample composition until they converge on a jointly consistent solution. As in many machine-learning problems, the algorithm must begin with an initial guess, for which the maximum possible error rates in this data are used (the error rates if only the most abundant sequence is correct and all the rest are errors).</p>
<div id="learn-the-error-rates" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Learn the Error Rates<a href="denoise.html#learn-the-error-rates" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Note: this is the most computationally intensive part of the process. Depending on your computer, this could take a while.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="denoise.html#cb1-1" aria-hidden="true" tabindex="-1"></a>errF <span class="ot">&lt;-</span> <span class="fu">learnErrors</span>(ffilt, <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 113049300 total bases in 434805 reads from 6 samples will be used for learning the error rates.</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="denoise.html#cb3-1" aria-hidden="true" tabindex="-1"></a>errR <span class="ot">&lt;-</span> <span class="fu">learnErrors</span>(rfilt, <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 110833580 total bases in 503789 reads from 7 samples will be used for learning the error rates.</code></pre>
<p>Visualize the estimated error rates:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="denoise.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotErrors</span>(errF, <span class="at">nominalQ =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
<p><img src="16Sworkbook_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>The error rates for each possible transition (A→C, A→G, …) are shown. Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score. Here the estimated error rates (black line) are a good fit to the observed rates (points), and the error rates drop with increased quality as expected. Everything looks reasonable and we proceed with confidence.</p>
</div>
<div id="dereplicate" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Dereplicate<a href="denoise.html#dereplicate" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="denoise.html#cb7-1" aria-hidden="true" tabindex="-1"></a>derep_forward <span class="ot">&lt;-</span> <span class="fu">derepFastq</span>(ffilt)</span>
<span id="cb7-2"><a href="denoise.html#cb7-2" aria-hidden="true" tabindex="-1"></a>derep_reverse <span class="ot">&lt;-</span> <span class="fu">derepFastq</span>(rfilt)</span></code></pre></div>
</div>
<div id="sample-inference" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> Sample Inference<a href="denoise.html#sample-inference" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We are now ready to apply the core sample inference algorithm to the filtered and trimmed sequence data.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="denoise.html#cb8-1" aria-hidden="true" tabindex="-1"></a>dadaFs <span class="ot">&lt;-</span> <span class="fu">dada</span>(derep_forward, <span class="at">err =</span> errF, <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Sample 1 - 66965 reads in 14499 unique sequences.
## Sample 2 - 68750 reads in 14269 unique sequences.
## Sample 3 - 68855 reads in 14142 unique sequences.
## Sample 4 - 80588 reads in 18364 unique sequences.
## Sample 5 - 61147 reads in 13962 unique sequences.
## Sample 6 - 88500 reads in 19081 unique sequences.
## Sample 7 - 68984 reads in 16567 unique sequences.
## Sample 8 - 95488 reads in 21378 unique sequences.
## Sample 9 - 55776 reads in 13093 unique sequences.
## Sample 10 - 61165 reads in 12912 unique sequences.
## Sample 11 - 55911 reads in 11642 unique sequences.
## Sample 12 - 67328 reads in 13655 unique sequences.
## Sample 13 - 55394 reads in 12090 unique sequences.
## Sample 14 - 78256 reads in 15288 unique sequences.
## Sample 15 - 82424 reads in 17243 unique sequences.
## Sample 16 - 61222 reads in 14529 unique sequences.
## Sample 17 - 50479 reads in 11478 unique sequences.
## Sample 18 - 61237 reads in 14215 unique sequences.
## Sample 19 - 67882 reads in 16006 unique sequences.
## Sample 20 - 67074 reads in 16233 unique sequences.
## Sample 21 - 51295 reads in 12722 unique sequences.
## Sample 22 - 68261 reads in 16104 unique sequences.
## Sample 23 - 67814 reads in 15683 unique sequences.
## Sample 24 - 71740 reads in 16231 unique sequences.</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="denoise.html#cb10-1" aria-hidden="true" tabindex="-1"></a>dadaRs <span class="ot">&lt;-</span> <span class="fu">dada</span>(derep_reverse, <span class="at">err =</span> errR, <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Sample 1 - 66965 reads in 16528 unique sequences.
## Sample 2 - 68750 reads in 16290 unique sequences.
## Sample 3 - 68855 reads in 16803 unique sequences.
## Sample 4 - 80588 reads in 21091 unique sequences.
## Sample 5 - 61147 reads in 15582 unique sequences.
## Sample 6 - 88500 reads in 21283 unique sequences.
## Sample 7 - 68984 reads in 18220 unique sequences.
## Sample 8 - 95488 reads in 24393 unique sequences.
## Sample 9 - 55776 reads in 15098 unique sequences.
## Sample 10 - 61165 reads in 14811 unique sequences.
## Sample 11 - 55911 reads in 14330 unique sequences.
## Sample 12 - 67328 reads in 17202 unique sequences.
## Sample 13 - 55394 reads in 13884 unique sequences.
## Sample 14 - 78256 reads in 17193 unique sequences.
## Sample 15 - 82424 reads in 17945 unique sequences.
## Sample 16 - 61222 reads in 19543 unique sequences.
## Sample 17 - 50479 reads in 13027 unique sequences.
## Sample 18 - 61237 reads in 14767 unique sequences.
## Sample 19 - 67882 reads in 18135 unique sequences.
## Sample 20 - 67074 reads in 17371 unique sequences.
## Sample 21 - 51295 reads in 14546 unique sequences.
## Sample 22 - 68261 reads in 18187 unique sequences.
## Sample 23 - 67814 reads in 17825 unique sequences.
## Sample 24 - 71740 reads in 18933 unique sequences.</code></pre>
<p>Inspecting the returned dada-class object:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="denoise.html#cb12-1" aria-hidden="true" tabindex="-1"></a>dadaFs[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## dada-class: object describing DADA2 denoising results
## 411 sequence variants were inferred from 14499 input unique sequences.
## Key parameters: OMEGA_A = 1e-40, OMEGA_C = 1e-40, BAND_SIZE = 16</code></pre>
<p>There is much more to the dada-class return object than this (see help(“dada-class”) for some info), including multiple diagnostics about the quality of each denoised sequence variant, but that is beyond the scope of an introductory tutorial.</p>
</div>
<div id="merge-reads" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> Merge reads<a href="denoise.html#merge-reads" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We now merge the forward and reverse reads together to obtain the full denoised sequences. Merging is performed by aligning the denoised forward reads with the reverse-complement of the corresponding denoised reverse reads, and then constructing the merged “contig” sequences. By default, merged sequences are only output if the forward and reverse reads overlap by at least 12 bases, and are identical to each other in the overlap region (but these conditions can be changed via function arguments).</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="denoise.html#cb14-1" aria-hidden="true" tabindex="-1"></a>merged <span class="ot">&lt;-</span> <span class="fu">mergePairs</span>(dadaFs, ffilt, dadaRs, rfilt, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 63415 paired-reads (in 621 unique pairings) successfully merged out of 65217 (in 1599 pairings) input.</code></pre>
<pre><code>## 65294 paired-reads (in 653 unique pairings) successfully merged out of 67067 (in 1548 pairings) input.</code></pre>
<pre><code>## 65507 paired-reads (in 676 unique pairings) successfully merged out of 67342 (in 1676 pairings) input.</code></pre>
<pre><code>## 75663 paired-reads (in 700 unique pairings) successfully merged out of 78383 (in 1997 pairings) input.</code></pre>
<pre><code>## 57711 paired-reads (in 610 unique pairings) successfully merged out of 59547 (in 1484 pairings) input.</code></pre>
<pre><code>## 83548 paired-reads (in 855 unique pairings) successfully merged out of 86413 (in 2293 pairings) input.</code></pre>
<pre><code>## 64830 paired-reads (in 662 unique pairings) successfully merged out of 67115 (in 1810 pairings) input.</code></pre>
<pre><code>## 90697 paired-reads (in 720 unique pairings) successfully merged out of 93089 (in 1993 pairings) input.</code></pre>
<pre><code>## 52380 paired-reads (in 509 unique pairings) successfully merged out of 54093 (in 1426 pairings) input.</code></pre>
<pre><code>## 58690 paired-reads (in 411 unique pairings) successfully merged out of 60020 (in 1044 pairings) input.</code></pre>
<pre><code>## 53132 paired-reads (in 515 unique pairings) successfully merged out of 54527 (in 1174 pairings) input.</code></pre>
<pre><code>## 64551 paired-reads (in 594 unique pairings) successfully merged out of 66050 (in 1290 pairings) input.</code></pre>
<pre><code>## 52686 paired-reads (in 513 unique pairings) successfully merged out of 54126 (in 1247 pairings) input.</code></pre>
<pre><code>## 74712 paired-reads (in 549 unique pairings) successfully merged out of 76912 (in 1390 pairings) input.</code></pre>
<pre><code>## 78767 paired-reads (in 645 unique pairings) successfully merged out of 81018 (in 1645 pairings) input.</code></pre>
<pre><code>## 57684 paired-reads (in 433 unique pairings) successfully merged out of 59412 (in 1285 pairings) input.</code></pre>
<pre><code>## 47437 paired-reads (in 519 unique pairings) successfully merged out of 48907 (in 1294 pairings) input.</code></pre>
<pre><code>## 57818 paired-reads (in 537 unique pairings) successfully merged out of 59462 (in 1410 pairings) input.</code></pre>
<pre><code>## 63513 paired-reads (in 646 unique pairings) successfully merged out of 65929 (in 1941 pairings) input.</code></pre>
<pre><code>## 62950 paired-reads (in 772 unique pairings) successfully merged out of 65102 (in 1937 pairings) input.</code></pre>
<pre><code>## 47747 paired-reads (in 558 unique pairings) successfully merged out of 49575 (in 1570 pairings) input.</code></pre>
<pre><code>## 64619 paired-reads (in 637 unique pairings) successfully merged out of 66388 (in 1578 pairings) input.</code></pre>
<pre><code>## 63995 paired-reads (in 565 unique pairings) successfully merged out of 66002 (in 1552 pairings) input.</code></pre>
<pre><code>## 68111 paired-reads (in 588 unique pairings) successfully merged out of 69645 (in 1428 pairings) input.</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="denoise.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the merger data.frame from the first sample</span></span>
<span id="cb39-2"><a href="denoise.html#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                                                                                                                                                                                                                                                                                                                                                                                                                                      sequence
## 1                          TGGGGAATATTGCACAATGGGGGAAACCCTGATGCAGCGACGCCGCGTGAAGGAAGAAGTATCTCGGTATGTAAACTTCTATCAGCAGGGAAGATAGTGACGGTACCTGACTAAGAAGCCCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGTGTGGCAAGTCTGATGTGAAAGGCATGGGCTCAACCTGTGGACTGCATTGGAAACTGTCATACTTGAGTGCCGGAGGGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACA
## 2                 TGGGGAATATTGCACAATGGGCGCAAGCCTGATGCAGCGACGCCGCGTGCGGGATGACGGCCTTCGGGTTGTAAACCGCTTTTGACTGGGAGCAAGCCCTTCGGGGTGAGTGTACCTTTCGAATAAGCACCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTATCCGGAATTATTGGGCGTAAAGGGCTCGTAGGCGGTTCGTCGCGTCCGGTGTGAAAGTCCATCGCTTAACGGTGGATCCGCGCCGGGTACGGGCGGGCTTGAGTGCGGTAGGGGAGACTGGAATTCCCGGTGTAACGGTGGAATGTGTAGATATCGGGAAGAACACCAATGGCGAAGGCAGGTCTCTGGGCCGTCACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACA
## 3 TGGGGAATCTTCCGCAATGGGCGAAAGCCTGACGGAGCAACGCCGCGTGAGTGATGACGGCCTTCGGGTTGTAAAACTCTGTGATCCGGGACGAAAAGGCAGAGTGCGAAGAACAAACTGCATTGACGGTACCGGAAAAGCAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGCTTCCCAAGTCCCTCTTAAAAGTGCGGGGCTTAACCCCGTGATGGGAAGGAAACTGGGAAGCTGGAGTATCGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCGGTGGCGAAGGCGACTTTCTGGACGAAAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACA
## 4                          TGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGCGAAGAAGTATTTCGGTATGTAAAGCTCTATCAGCAGGGAAGATAATGACGGTACCTGACTAAGAAGCACCGGCTAAATACGTGCCAGCAGCCGCGGTAATACGTATGGTGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGCAGGCGGTGCGGCAAGTCTGATGTGAAAGCCCGGGGCTCAACCCCGGTACTGCATTGGAAACTGTCGTACTAGAGTGTCGGAGGGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGATAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACA
## 5                          TGGGGAATATTGCACAATGGGGGAAACCCTGATGCAGCGACGCCGCGTGGAGGAAGAAGGTCTTCGGATTGTAAACTCCTGTTGTTGAGGAAGATAATGACGGTACTCAACAAGGAAGTGACGGCTAACTACGTGCCAGCAGCCGCGGTAAAACGTAGGTCACAAGCGTTGTCCGGAATTACTGGGTGTAAAGGGAGCGCAGGCGGGCGATCAAGTTGGAAGTGAAATCCATGGGCTCAACCCATGAACTGCTTTCAAAACTGGTCGTCTTGAGTAGTGCAGAGGTAGGCGGAATTCCCGGTGTAGCGGTGGAATGCGTAGATATCGGGAGGAACACCAGTGGCGAAGGCGGCCTACTGGGCACCAACTGACGCTGAGGCTCGAAAGTGTGGGTAGCAAACA
## 6                         TGGGGAATCTTGCGCAATGGGGGGAACCCTGACGCAGCGACGCCGCGTGCGGGACGGAGGCCTTCGGGTCGTAAACCGCTTTCAGCAGGGAAGAGTCAAGACTGTACCTGCAGAAGAAGCCCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGGGCGAGCGTTATCCGGATTCATTGGGCGTAAAGCGCGCGTAGGCGGCCCGGCAGGCCGGGGGTCGAAGCGGGGGGCTCAACCCCCCGAAGCCCCCGGAACCTCCGCGGCTTGGGTCCGGTAGGGGAGGGTGGAACACCCGGTGTAGCGGTGGAATGCGCAGATATCGGGTGGAACACCGGTGGCGAAGGCGGCCCTCTGGGCCGAGACCGACGCTGAGGCGCGAAAGCTGGGGGAGCGAACA
##   abundance forward reverse nmatch nmismatch nindel prefer accept
## 1      4464       1       1     78         0      0      1   TRUE
## 2      4074       3       3     69         0      0      1   TRUE
## 3      4067       2       5     53         0      0      1   TRUE
## 4      3150       4       2     78         0      0      2   TRUE
## 5      3134       5       4     78         0      0      1   TRUE
## 6      2800       6       8     77         0      0      1   TRUE</code></pre>
<p>The <code>merged</code> object is a list of data.frames from each sample. Each data.frame contains the merged <code>$sequence</code>, its <code>$abundance</code>, and the indices of the <code>$forward</code> and <code>$reverse</code> sequence variants that were merged. Paired reads that did not exactly overlap were removed by <code>mergePairs</code>, further reducing spurious output.</p>
<p><strong>Considerations</strong>: Most of your reads should successully merge.If that is not the case upstream parameters may need to be revisited: Did you trim away the overlap between your reads?</p>
</div>
<div id="construct-sequence-table" class="section level2 hasAnchor" number="5.5">
<h2><span class="header-section-number">5.5</span> Construct sequence table<a href="denoise.html#construct-sequence-table" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We can now construct an amplicon sequence variant table (ASV) table, a higher-resolution version of the OTU table produced by traditional methods.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="denoise.html#cb41-1" aria-hidden="true" tabindex="-1"></a>seqtab <span class="ot">&lt;-</span> <span class="fu">makeSequenceTable</span>(merged)</span>
<span id="cb41-2"><a href="denoise.html#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(seqtab)</span></code></pre></div>
<pre><code>## [1]   24 7656</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="denoise.html#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect distribution of sequence lengths</span></span>
<span id="cb43-2"><a href="denoise.html#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">nchar</span>(<span class="fu">getSequences</span>(seqtab)))</span></code></pre></div>
<pre><code>## 
##  260  282  300  315  323  365  366  372  373  377  380  385  386  393  400  401 
##    4    1    1    1    2    1    1    1    2    1    1    2    1    1    3  372 
##  402  403  404  405  406  407  408  409  410  411  412  416  418  419  420  421 
## 3686  744  300  429   31  159   76  160   31  141    3    1    2    3    7   62 
##  422  423  424  425  426  427  428  429  434  439  463 
##  536   15    1    3  198  484  185    1    1    1    1</code></pre>
<p>The sequence table is a matrix with rows corresponding to (and named by) the samples, and columns corresponding to (and named by) the sequence variants.</p>
<div id="exercise-4" class="section level3 hasAnchor" number="5.5.1">
<h3><span class="header-section-number">5.5.1</span> Exercise<a href="denoise.html#exercise-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Are these sequences the length that you expect? How are the reads concentrated in the merged sequence lengths? Plot the distribution of sequence lengths to explore.</p>
</div>
<div id="exercise-5" class="section level3 hasAnchor" number="5.5.2">
<h3><span class="header-section-number">5.5.2</span> Exercise<a href="denoise.html#exercise-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Sequences that are much longer or shorter than expected may be the result of non-specific priming. Remove the non-target-length sequences from your sequence table. This is analogous to “cutting a band” in-silico to get amplicons of the targeted length.</p>
<p><strong>Hint</strong>. You may find the function <code>nchar</code> useful.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="filter.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["16Sworkbook.pdf", "16Sworkbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
