<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Phyloseq | 16s Microbiome Analysis Workshop</title>
  <meta name="description" content="This workbook contains workshop material for all three days." />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Phyloseq | 16s Microbiome Analysis Workshop" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This workbook contains workshop material for all three days." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Phyloseq | 16s Microbiome Analysis Workshop" />
  
  <meta name="twitter:description" content="This workbook contains workshop material for all three days." />
  

<meta name="author" content="Alana Schick" />


<meta name="date" content="2022-08-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="assign-taxonomy.html"/>
<link rel="next" href="visualization.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">16S Analysis Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#workshop-schedule"><i class="fa fa-check"></i>Workshop Schedule</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-1"><i class="fa fa-check"></i>Day 1</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-2"><i class="fa fa-check"></i>Day 2</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-3"><i class="fa fa-check"></i>Day 3</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#important-links"><i class="fa fa-check"></i>Important links</a></li>
</ul></li>
<li class="part"><span><b>I Day 1</b></span></li>
<li class="chapter" data-level="1" data-path="pre-workshop.html"><a href="pre-workshop.html"><i class="fa fa-check"></i><b>1</b> Pre-workshop</a>
<ul>
<li class="chapter" data-level="1.1" data-path="pre-workshop.html"><a href="pre-workshop.html#packages"><i class="fa fa-check"></i><b>1.1</b> Packages</a></li>
<li class="chapter" data-level="1.2" data-path="pre-workshop.html"><a href="pre-workshop.html#data-download"><i class="fa fa-check"></i><b>1.2</b> Data download</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="introduction.html"><a href="introduction.html#basics-and-background"><i class="fa fa-check"></i><b>2.1</b> Basics and Background</a></li>
<li class="chapter" data-level="2.2" data-path="introduction.html"><a href="introduction.html#data"><i class="fa fa-check"></i><b>2.2</b> Data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="introduction.html"><a href="introduction.html#exercise"><i class="fa fa-check"></i><b>2.2.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="removing-primers.html"><a href="removing-primers.html"><i class="fa fa-check"></i><b>3</b> Removing Primers</a>
<ul>
<li class="chapter" data-level="3.1" data-path="removing-primers.html"><a href="removing-primers.html#identify-primers"><i class="fa fa-check"></i><b>3.1</b> Identify Primers</a></li>
<li class="chapter" data-level="3.2" data-path="removing-primers.html"><a href="removing-primers.html#remove-primers"><i class="fa fa-check"></i><b>3.2</b> Remove Primers</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="filter.html"><a href="filter.html"><i class="fa fa-check"></i><b>4</b> Filter</a>
<ul>
<li class="chapter" data-level="4.1" data-path="filter.html"><a href="filter.html#inspect"><i class="fa fa-check"></i><b>4.1</b> Inspect</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="filter.html"><a href="filter.html#exercise-1"><i class="fa fa-check"></i><b>4.1.1</b> Exercise</a></li>
<li class="chapter" data-level="4.1.2" data-path="filter.html"><a href="filter.html#exercise-2"><i class="fa fa-check"></i><b>4.1.2</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="filter.html"><a href="filter.html#filter-and-trim"><i class="fa fa-check"></i><b>4.2</b> Filter and Trim</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="filter.html"><a href="filter.html#exercise-3"><i class="fa fa-check"></i><b>4.2.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="filter.html"><a href="filter.html#exploring-parameters"><i class="fa fa-check"></i><b>4.3</b> Exploring parameters</a></li>
</ul></li>
<li class="part"><span><b>II Day 2</b></span></li>
<li class="chapter" data-level="5" data-path="denoise.html"><a href="denoise.html"><i class="fa fa-check"></i><b>5</b> Denoise</a>
<ul>
<li class="chapter" data-level="5.1" data-path="denoise.html"><a href="denoise.html#learn-the-error-rates"><i class="fa fa-check"></i><b>5.1</b> Learn the Error Rates</a></li>
<li class="chapter" data-level="5.2" data-path="denoise.html"><a href="denoise.html#dereplicate"><i class="fa fa-check"></i><b>5.2</b> Dereplicate</a></li>
<li class="chapter" data-level="5.3" data-path="denoise.html"><a href="denoise.html#sample-inference"><i class="fa fa-check"></i><b>5.3</b> Sample Inference</a></li>
<li class="chapter" data-level="5.4" data-path="denoise.html"><a href="denoise.html#merge-reads"><i class="fa fa-check"></i><b>5.4</b> Merge reads</a></li>
<li class="chapter" data-level="5.5" data-path="denoise.html"><a href="denoise.html#construct-sequence-table"><i class="fa fa-check"></i><b>5.5</b> Construct sequence table</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="denoise.html"><a href="denoise.html#exercise-4"><i class="fa fa-check"></i><b>5.5.1</b> Exercise</a></li>
<li class="chapter" data-level="5.5.2" data-path="denoise.html"><a href="denoise.html#exercise-5"><i class="fa fa-check"></i><b>5.5.2</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="denoise.html"><a href="denoise.html#remove-chimeras"><i class="fa fa-check"></i><b>5.6</b> Remove Chimeras</a></li>
<li class="chapter" data-level="5.7" data-path="denoise.html"><a href="denoise.html#track-reads-through-the-pipeline"><i class="fa fa-check"></i><b>5.7</b> Track reads through the pipeline</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="denoise.html"><a href="denoise.html#exercise-6"><i class="fa fa-check"></i><b>5.7.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="assign-taxonomy.html"><a href="assign-taxonomy.html"><i class="fa fa-check"></i><b>6</b> Assign Taxonomy</a>
<ul>
<li class="chapter" data-level="6.1" data-path="assign-taxonomy.html"><a href="assign-taxonomy.html#download-the-database"><i class="fa fa-check"></i><b>6.1</b> Download the database</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="assign-taxonomy.html"><a href="assign-taxonomy.html#exercise-7"><i class="fa fa-check"></i><b>6.1.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Day 3</b></span></li>
<li class="chapter" data-level="7" data-path="phyloseq.html"><a href="phyloseq.html"><i class="fa fa-check"></i><b>7</b> Phyloseq</a>
<ul>
<li class="chapter" data-level="7.1" data-path="phyloseq.html"><a href="phyloseq.html#import"><i class="fa fa-check"></i><b>7.1</b> Import</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="phyloseq.html"><a href="phyloseq.html#exercise-8"><i class="fa fa-check"></i><b>7.1.1</b> Exercise</a></li>
<li class="chapter" data-level="7.1.2" data-path="phyloseq.html"><a href="phyloseq.html#exercise-9"><i class="fa fa-check"></i><b>7.1.2</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="phyloseq.html"><a href="phyloseq.html#preprocess"><i class="fa fa-check"></i><b>7.2</b> Preprocess</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="phyloseq.html"><a href="phyloseq.html#reads-per-sample"><i class="fa fa-check"></i><b>7.2.1</b> Reads per sample</a></li>
<li class="chapter" data-level="7.2.2" data-path="phyloseq.html"><a href="phyloseq.html#exercise-10"><i class="fa fa-check"></i><b>7.2.2</b> Exercise</a></li>
<li class="chapter" data-level="7.2.3" data-path="phyloseq.html"><a href="phyloseq.html#exercise-11"><i class="fa fa-check"></i><b>7.2.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="phyloseq.html"><a href="phyloseq.html#filtering"><i class="fa fa-check"></i><b>7.3</b> Filtering</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="phyloseq.html"><a href="phyloseq.html#exercise-12"><i class="fa fa-check"></i><b>7.3.1</b> Exercise</a></li>
<li class="chapter" data-level="7.3.2" data-path="phyloseq.html"><a href="phyloseq.html#outlier-detection"><i class="fa fa-check"></i><b>7.3.2</b> Outlier Detection</a></li>
<li class="chapter" data-level="7.3.3" data-path="phyloseq.html"><a href="phyloseq.html#exercise-13"><i class="fa fa-check"></i><b>7.3.3</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="visualization.html"><a href="visualization.html"><i class="fa fa-check"></i><b>8</b> Visualization</a>
<ul>
<li class="chapter" data-level="8.1" data-path="visualization.html"><a href="visualization.html#taxonomic-composition"><i class="fa fa-check"></i><b>8.1</b> Taxonomic Composition</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="visualization.html"><a href="visualization.html#exercise-14"><i class="fa fa-check"></i><b>8.1.1</b> Exercise</a></li>
<li class="chapter" data-level="8.1.2" data-path="visualization.html"><a href="visualization.html#exercise-15"><i class="fa fa-check"></i><b>8.1.2</b> Exercise</a></li>
<li class="chapter" data-level="8.1.3" data-path="visualization.html"><a href="visualization.html#exercise-16"><i class="fa fa-check"></i><b>8.1.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="visualization.html"><a href="visualization.html#alpha-diversity"><i class="fa fa-check"></i><b>8.2</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="visualization.html"><a href="visualization.html#exercise-17"><i class="fa fa-check"></i><b>8.2.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="visualization.html"><a href="visualization.html#beta-diversity"><i class="fa fa-check"></i><b>8.3</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="visualization.html"><a href="visualization.html#exercise-18"><i class="fa fa-check"></i><b>8.3.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">16s Microbiome Analysis Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="phyloseq" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">7</span> Phyloseq<a href="phyloseq.html#phyloseq" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The phyloseq R package is a powerful framework for further analysis of microbiome data. We now demonstrate how to straightforwardly import the tables produced by the <code>dada2</code> pipeline into phyloseq.</p>
<div id="import" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Import<a href="phyloseq.html#import" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="phyloseq.html#cb1-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">&quot;~/Desktop/microbiomeworkshop2022/&quot;</span></span>
<span id="cb1-2"><a href="phyloseq.html#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="phyloseq.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Read in files</span></span>
<span id="cb1-4"><a href="phyloseq.html#cb1-4" aria-hidden="true" tabindex="-1"></a>seqtab <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path, <span class="st">&quot;results/seqtab_final.rds&quot;</span>))</span>
<span id="cb1-5"><a href="phyloseq.html#cb1-5" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path, <span class="st">&quot;results/taxa_final.rds&quot;</span>))</span></code></pre></div>
<hr />
<div id="exercise-8" class="section level3 hasAnchor" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Exercise<a href="phyloseq.html#exercise-8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li><p>Read the metadata file into R, called <code>info</code>.</p></li>
<li><p>Phyloseq needs the rownames of the metadata to match the sample names. How can you accomplish this?</p></li>
</ol>
<hr />
<p>Before creating your phyloseq object, familiarize yourself with the metadata. Are there any other changes you need to make?</p>
<p>It can sometimes be useful to clean up the sample names at this point - Illumina will add their own sample numbers to your sample names. You can do this by creating a new variable that doesnâ€™t contain the information Illumina added:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="phyloseq.html#cb2-1" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> info <span class="sc">%&gt;%</span> <span class="fu">separate</span>(SampleID, <span class="fu">c</span>(<span class="st">&quot;SampleID&quot;</span>, <span class="st">&quot;temp&quot;</span>), <span class="at">sep =</span> <span class="st">&quot;_S&quot;</span>)</span></code></pre></div>
<p>Make a phyloseq object:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="phyloseq.html#cb3-1" aria-hidden="true" tabindex="-1"></a>ps <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(<span class="fu">otu_table</span>(seqtab, <span class="at">taxa_are_rows=</span><span class="cn">FALSE</span>), <span class="fu">sample_data</span>(info), <span class="fu">tax_table</span>(taxa))</span>
<span id="cb3-2"><a href="phyloseq.html#cb3-2" aria-hidden="true" tabindex="-1"></a>ps</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 1201 taxa and 24 samples ]
## sample_data() Sample Data:       [ 24 samples by 5 sample variables ]
## tax_table()   Taxonomy Table:    [ 1201 taxa by 7 taxonomic ranks ]</code></pre>
<p>In order to look at the ASVs, we need to remove the sequences.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="phyloseq.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Remove sequence names, rename to something manageable</span></span>
<span id="cb5-2"><a href="phyloseq.html#cb5-2" aria-hidden="true" tabindex="-1"></a>asv_names <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="fu">dim</span>(<span class="fu">otu_table</span>(ps))[<span class="dv">2</span>], <span class="at">mode =</span> <span class="st">&quot;character&quot;</span>)</span>
<span id="cb5-3"><a href="phyloseq.html#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(<span class="fu">otu_table</span>(ps))[<span class="dv">2</span>]){</span>
<span id="cb5-4"><a href="phyloseq.html#cb5-4" aria-hidden="true" tabindex="-1"></a>    asv_names[i] <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;ASV&quot;</span>, i, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb5-5"><a href="phyloseq.html#cb5-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-6"><a href="phyloseq.html#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">taxa_names</span>(ps) <span class="ot">&lt;-</span> asv_names</span>
<span id="cb5-7"><a href="phyloseq.html#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">otu_table</span>(ps)) <span class="ot">&lt;-</span> asv_names</span>
<span id="cb5-8"><a href="phyloseq.html#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(<span class="fu">tax_table</span>(ps)) <span class="ot">&lt;-</span> asv_names</span></code></pre></div>
<p>You can see that the phyloseq object has an <code>otu_table</code>(ASV table), <code>sample_data</code> and <code>tax_table</code>. You can use functions <code>tax_table()</code>, <code>sample_data()</code> and <code>otu_table()</code> to access the data.</p>
<hr />
</div>
<div id="exercise-9" class="section level3 hasAnchor" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Exercise<a href="phyloseq.html#exercise-9" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Take a look the following functions and find out what they do:</p>
<ul>
<li>subset_samples()</li>
<li>subset_taxa()</li>
<li>tax_glom()</li>
<li>sample_sums()</li>
<li>prune_samples()</li>
<li>transform_sample_counts()</li>
<li>psmelt()</li>
</ul>
<hr />
</div>
</div>
<div id="preprocess" class="section level2 hasAnchor" number="7.2">
<h2><span class="header-section-number">7.2</span> Preprocess<a href="phyloseq.html#preprocess" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="reads-per-sample" class="section level3 hasAnchor" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> Reads per sample<a href="phyloseq.html#reads-per-sample" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In general, the first step in pre-processing is to check how many reads you have per sample and remove any samples if they failed.</p>
<hr />
</div>
<div id="exercise-10" class="section level3 hasAnchor" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Exercise<a href="phyloseq.html#exercise-10" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Use either the <code>rowSums()</code> or <code>colSums()</code> function (or any other function of your choosing) to create a variable that contains the total number of reads per sample.</p>
<hr />
<p>Once you have this, create a data frame containing this information:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="phyloseq.html#cb6-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">as</span>(<span class="fu">sample_data</span>(ps), <span class="st">&quot;data.frame&quot;</span>), <span class="at">TotalReads =</span> sums)</span>
<span id="cb6-2"><a href="phyloseq.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(counts)</span></code></pre></div>
<pre><code>##                SampleID   temp SubjectID Treatment Timepoint TotalReads
## 01MP-1_S1_L001   01MP-1 1_L001      PS25         A         1      64763
## 01MP-2_S2_L001   01MP-2 2_L001      PS25         A         2      66378
## 01MP-3_S3_L001   01MP-3 3_L001      PS25         A         3      65698
## 03KW-1_S4_L001   03KW-1 4_L001      PS19         A         1      81586
## 03KW-2_S5_L001   03KW-2 5_L001      PS19         A         2      55418
## 03KW-3_S6_L001   03KW-3 6_L001      PS19         A         3      81770</code></pre>
<hr />
</div>
<div id="exercise-11" class="section level3 hasAnchor" number="7.2.3">
<h3><span class="header-section-number">7.2.3</span> Exercise<a href="phyloseq.html#exercise-11" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Create a visualization of these results (something like the one below) and decide if any of the samples need to be removed.</p>
<p><img src="16Sworkbook_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<hr />
<p><strong>Other considerations</strong>: If there is a large amount of variation in the number of reads across samples (in general, more than 10-fold), you need to take steps to normalize the data.</p>
</div>
</div>
<div id="filtering" class="section level2 hasAnchor" number="7.3">
<h2><span class="header-section-number">7.3</span> Filtering<a href="phyloseq.html#filtering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here, we filter out ASVs (amplicon sequence variants) using two criteria: <strong>abundance</strong> and <strong>prevalence</strong>. First, compute the prevalence of each ASV by defining prevalence as the number of samples in which a taxon appears at least once:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="phyloseq.html#cb8-1" aria-hidden="true" tabindex="-1"></a>prevdf <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="at">X =</span> <span class="fu">otu_table</span>(ps), <span class="at">MARGIN =</span> <span class="fu">ifelse</span>(<span class="fu">taxa_are_rows</span>(ps), <span class="at">yes =</span> <span class="dv">1</span>, <span class="at">no =</span> <span class="dv">2</span>), <span class="at">FUN =</span> <span class="cf">function</span>(x){<span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>)})</span>
<span id="cb8-2"><a href="phyloseq.html#cb8-2" aria-hidden="true" tabindex="-1"></a>prevdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">prevalence =</span> prevdf, <span class="at">total_abundance =</span> <span class="fu">taxa_sums</span>(ps), <span class="fu">tax_table</span>(ps))</span></code></pre></div>
<p>Visualize prevalence:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="phyloseq.html#cb9-1" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(prevdf, <span class="fu">aes</span>(total_abundance, prevalence<span class="sc">/</span><span class="fu">nsamples</span>(ps), <span class="at">colour =</span> Phylum)) <span class="sc">+</span> </span>
<span id="cb9-2"><a href="phyloseq.html#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb9-3"><a href="phyloseq.html#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Total abundance&quot;</span>) <span class="sc">+</span> </span>
<span id="cb9-4"><a href="phyloseq.html#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Prevalence (fraction samples)&quot;</span>) <span class="sc">+</span> </span>
<span id="cb9-5"><a href="phyloseq.html#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb9-6"><a href="phyloseq.html#cb9-6" aria-hidden="true" tabindex="-1"></a>gg</span></code></pre></div>
<p><img src="16Sworkbook_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<hr />
<div id="exercise-12" class="section level3 hasAnchor" number="7.3.1">
<h3><span class="header-section-number">7.3.1</span> Exercise<a href="phyloseq.html#exercise-12" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Use this visualization to determine your filtering parameters and define them as <code>prevalence_threshold</code> and <code>count_threshold</code>. Add these parameters to your plot using the functions <code>geom_hline()</code> and <code>geom_vline()</code>.</p>
<p><img src="16Sworkbook_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<hr />
<p>Define taxa to filter:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="phyloseq.html#cb10-1" aria-hidden="true" tabindex="-1"></a>keeptaxa <span class="ot">&lt;-</span> <span class="fu">rownames</span>(prevdf)[(prevdf<span class="sc">$</span>prevalence <span class="sc">&gt;</span> prevalence_threshold) <span class="sc">&amp;</span> (prevdf<span class="sc">$</span>total_abundance <span class="sc">&gt;</span> count_threshold)]</span></code></pre></div>
<p>Execute filter:</p>
<p>Important to keep your unaltered data!</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="phyloseq.html#cb11-1" aria-hidden="true" tabindex="-1"></a>psf <span class="ot">&lt;-</span> <span class="fu">prune_taxa</span>(keeptaxa, ps)</span></code></pre></div>
<p>Compute relative abundance (of both raw and filtered data):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="phyloseq.html#cb12-1" aria-hidden="true" tabindex="-1"></a>rel <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(ps, <span class="cf">function</span>(x) x <span class="sc">/</span> <span class="fu">sum</span>(x))</span>
<span id="cb12-2"><a href="phyloseq.html#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="phyloseq.html#cb12-3" aria-hidden="true" tabindex="-1"></a>relf <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(psf, <span class="cf">function</span>(x) x <span class="sc">/</span> <span class="fu">sum</span>(x))</span></code></pre></div>
</div>
<div id="outlier-detection" class="section level3 hasAnchor" number="7.3.2">
<h3><span class="header-section-number">7.3.2</span> Outlier Detection<a href="phyloseq.html#outlier-detection" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, log transform your count data:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="phyloseq.html#cb13-1" aria-hidden="true" tabindex="-1"></a>pslog <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(psf, <span class="cf">function</span>(x) <span class="fu">log</span>(<span class="dv">1</span> <span class="sc">+</span> x))</span></code></pre></div>
<p>Ordinate and get the eigenvalues for axes dimensions:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="phyloseq.html#cb14-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(pslog, <span class="at">method =</span> <span class="st">&quot;PCoA&quot;</span>, <span class="at">distance =</span> <span class="st">&quot;bray&quot;</span>)</span></code></pre></div>
<hr />
</div>
<div id="exercise-13" class="section level3 hasAnchor" number="7.3.3">
<h3><span class="header-section-number">7.3.3</span> Exercise<a href="phyloseq.html#exercise-13" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Create an ordination plot to check for any obvious outliers. Use the function <code>plot_ordination()</code> to do this.</p>
<p><img src="16Sworkbook_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>Do you think any of these samples are outliers? How could you investigate further?</p>
<hr />

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="assign-taxonomy.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="visualization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["16Sworkbook.pdf", "16Sworkbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
